### ISSUE TITLE
[CMake] Missing a way to define CMake args from the meson call

### ISSUE THREAD

**Issue by Neumann-A** [mesonbuild/meson#7223]
To make meson work correctly within vcpkg using cmake dependencies, meson needs to pass two things to CMake via meson:
- a cmake toolchain file
- a variable called VCPKG_TARGET_TRIPLET set to the correct value to use. 

As such having the option to globally pass `cmake_args` is requiered.

Furthermore using `--find-package` mode is deprecated. 
https://cmake.org/cmake/help/latest/manual/cmake.1.html#run-the-find-package-tool

**Comment by mensinda** [comment]
Just to clarify, this is about CMake dependencies and not CMake subprojects, right?

> a variable called VCPKG_TARGET_TRIPLET set to the correct value to use.

wouldn't this be redundant with the CMake toolchain file? Can't you just set this variable there or is this something highly specific?

> Furthermore using `--find-package` mode is deprecated.

I am aware and we are not using it (long and complicated storry involving the CMake trace).

**Comment by Neumann-A** [comment]
>Just to clarify, this is about CMake dependencies and not CMake subprojects, right?

Yes CMake dependencies

> wouldn't this be redundant with the CMake toolchain file? Can't you just set this variable there or is this something highly specific?

The toolchain is https://github.com/microsoft/vcpkg/blob/master/scripts/buildsystems/vcpkg.cmake
and `VCPKG_TARGET_TRIPLET` is the triplet selection variable. There is also `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` which can also be passed by users containing the 'real' toolchain specifying compiler flags and setup for cmake.
Furthermore `CMAKE_FIND_LIBRARY_(PREFIXES|SUFFIXES)` need to be adjusted due to meson's decision to name native static windows library like `lib<name>.a` which not only breaks `find_library` in cmake but also autotools with pkg-config on native windows.

Just let the user pass an arbitrary number of flags to the cmake invocation. Meson can simply not predict what the user might need to pass to cmake to make it work as the user wishes.

**Comment by mensinda** [comment]
> Just let the user pass an arbitrary number of flags to the cmake invocation.

In principle I have nothing against that, but not from the commandline. I would much rather use mesons cross/native files to set them there.

> Furthermore CMAKE_FIND_LIBRARY_(PREFIXES|SUFFIXES) need to be adjusted due to meson's decision to name native static windows library like lib<name>.a which not only breaks find_library in cmake but also autotools with pkg-config on native windows.

Why is this a problem here? Meson does not override anything regarding library suffixes for when invoking CMake. Or what is the usecase here.

**Comment by Neumann-A** [comment]
> In principle I have nothing against that, but not from the commandline. I would much rather use mesons cross/native files to set them there.

Yeah make my life complicated by first having me configure a file with the correct paths  (making file IO yeah....) and than passing that as file to meson instead of just adding directly `-Dcmake_args=['arg1','arg2','arg3']` to the cmd line call. 

> Why is this a problem here? Meson does not override anything regarding library suffixes for when invoking CMake. Or what is the usecase here.

That is the problem: Assume zlib switches to meson (which probably never happens). On windows static libraries of zlib will be called: `libzlib.a` breaking all consumers using other buildsystems. Not even meson's own generated pkg-config file for zlib would work because it would only contain `-lzlib` and not `-llibzlib` on windows. So meson itself must already modify `CMAKE_FIND_LIBRARY_(PREFIXES|SUFFIXES)` on windows to actually find libraries generated by itself.

**Comment by mensinda** [comment]
> Yeah make my life complicated by first having me...

It's about the user experience for everyone, not just vcpkg :) There is also #6597, which would allow you to do everything with cross/native files.

I don't think that a normal (as in setting similar use case, but doing things manually) user would want to specify 10 CMake variables over the command line, for every project. The cross/native files are just a lot nicer for something like this. Also, we like to keep the options list as short as possible.

> Not even meson's own generated pkg-config file for zlib would work because it would only contain -lzlib and not -llibzlib on windows.

I am not sure if this is intentional and I think this should be addressed. However, not in this issue (so we don't get too much offtopic). Could you please open a new issue with more details (expected behavior, actual behavior) so that this can be discussed/fixed?

**Comment by Neumann-A** [comment]
> I don't think that a normal (as in setting similar use case, but doing things manually) user would want to specify 10 CMake variables over the command line,

...... who said something about manual...... I am just saying for a build script invoking meson with different parameters I have to create/configure a (native) file and then pass that to meson. That just seems slow to me and more indirect than directly passing the required parameters. 

>Could you please open a new issue with more details (expected behavior, actual behavior) so that this can be discussed/fixed?

No (due to personal opinion about meson), only working with it to make it minimally correctly working with vcpkg which is difficult/impossible(?) due to all the limitations/design decisions within meson

**Comment by mensinda** [comment]
> ...... who said something about manual...... I am just saying for a build script invoking meson with different parameters I have to create/configure a (native) file and then pass that to meson.

My point was that I am not going to implement something that is just designed for vcpkg, but something that should work well for **all** end users (especially when adding a new feature). And to me at least, it currently looks like that implementing this with our cross/native files is the best solution for most users and I would categorize generating a native file only as mildly inconvenient.

I agree that it makes your job a little harder, but I definitely want to support setting CMake paramenters/defines inside a cross-file, since this is necessary to get cross-compilation with CMake subprojects going. Also, adding support for a command-line option since this would massively complicate the rules for CMake options (which args are appended, detecting CMake toolchain files and correctly handling them in the future when we have cross-compilation support for CMake dependencies/subprojects, etc).

As for library naming: boost also has a `lib` prefix for its static libraries. See [the boost docs](https://www.boost.org/doc/libs/1_73_0/more/getting_started/windows.html#library-naming).

This is necessary, because of the also static import libraries for each `*.dll`, which only differ in the prefix. If meson wouldn't also do this, we would have no way to implement our [`both_libraries`](https://mesonbuild.com/Reference-manual.html#both_libraries) function.

If mesons pkg-config module produces the wrong flags, then this is a bug and must be addressed.

> No (due to personal opinion about meson), only working with it to make it minimally...

Everyone is allowed to have his opinions :)

> ...with vcpkg which is difficult/impossible(?) due to all the limitations/design decisions within meson

You know, we are trying our best to be compatible with other software (introspection API, etc.). We don't want to make your life hard without a good reason (and yes specifying arbitrary CMake parameters will be implemented at some point, probably together with cross-compilation support).

So, telling us what the problems on mesons side with a possible vcpkg integration are would benefit both projects.

You are free to disagree with our design decisions, but we *do* have a design process and we are not adding features randomly.

**Comment by Neumann-A** [comment]
>complicate the rules for CMake options

I did not want anything complicated just a transparent forwarding of arguments to CMake would be enough. No analyses required, just stupid forwarding in the same order as the given python array. 

>This is necessary, because of the also static import libraries for each *.dll, which only differ in the prefix. If meson wouldn't also do this, we would have no way to implement our both_libraries function.

vcpkg always only builds one library (static or shared) so the feature to have `both_libraries` does not apply. I know the "story" behind the lib prefix and .a suffix and the "name" collision and still find the decision wrong since meson does not give the user a way to opt-out of that behavior. In CMake nearly everything can be user customized (without touching the build recipe[CMakeLists.txt]). In meson nearly everything is hardcoded and you need to change the meson.build or even meson itself.
I mean properly used autotools/make scripts are easier to handle than meson..... and I know since I worked on both in vcpkg. 

>we do have a design process

the problem here is that theprocess is strongly influenced by unix systems with all its assumptions.

**Comment by mensinda** [comment]
> No analyses required, just stupid forwarding in the same order as the given python array.

My main issue is that stupid forwarding will likely conflict with the upcoming cross-compiling support. Especially in regards to the toolchain files, which I intend to generate automatically based on the cross/native file.

As far as I am aware, there is currently no way to change the behavior of `both_libraries` to only build one library, even if the end-users want it. As for there being no way to set the library prefix/suffix (and other wings you want to be configurable): Something like this could be discussed in more detail in a different issue.

**Comment by Neumann-A** [comment]
>My main issue is that stupid forwarding will likely conflict with the upcoming cross-compiling support.

Oh no..... meson tries to be too clever. If the user provided a toolchain the user has to be in control.... 

>Especially in regards to the toolchain files, which I intend to generate automatically based on the cross/native file.

If you do that provide a way to generate the toolchain only without the requirement to have a meson project and if the user already provided a toolchain do nothing. This way the toolchain can be used without meson in cmake or it can be included in a user provided one.

**Comment by mensinda** [comment]
I do plan to support custom toolchain files by `include()`ing them at the end (so they can override the automatically generated stuff) and adding an option for only using the user-provided one only as a last resort.

I would prefer/recommend to `include()` the other toolchain file since we need to inject some CMake code to make certain features work.

> Oh no..... meson tries to be too clever.

Sarcasm doesn't really help to get your point across. Well-formed arguments will, however.

Specifying all this with command-line options will add quite a few options and will be difficult to understand for the user.

**Comment by Neumann-A** [comment]
>I do plan to support custom toolchain files by include()ing them at the end (so they can override the automatically generated stuff) and adding an option for only using the user-provided one only as a last resort.

I'm just speaking about finding dependencies nothing more. If you inject more there might be a problem especially if you are going to overwrite `add_library` or something like that.
As always give the user a way to opt-out of that behavior. She/He might be knowing more than the buildsystem in use. (Again keep the user in control instead of going into auto steer mode and drive into a hardcoded wall)

> Specifying all this with command-line options will add quite a few options and will be difficult to understand for the user.

Only one option..... `-Dcmake_args=['-Dcmakearg1','-Dcmakearg2',....]`. Your dependencies already have an option for `cmake_args`. I just want a global option.... if you want to copy CMake further call it `meson_cmake_args` similar to how cmake prefix some of its global initialization settings (e.g. [RUNTIME_OUTPUT_DIRECTORY](https://cmake.org/cmake/help/latest/variable/CMAKE_RUNTIME_OUTPUT_DIRECTORY.html) )

**Comment by mensinda** [comment]
As I have no pointed out multiple times already: I don't want users to just specify a toolchain file by adding random parameters to the CMake command-line. Doing so would require intransparent logic inside meson to do something else with the automatically generated toolchain files (disable them, `include()` the user one?), which may silently disable other features and cause errors. This is just bad UI, so I want to make setting a user toolchain more explicit with more configuration options.

I guess I could add a `-Dcmake_defines=['-DWHATEVER=ON']` meson option that only accepts elements starting with `-D`. This way, you could have a fixed meson cross/native file in vcpkg which specifies your toolchain file.

> I'm just speaking about finding dependencies nothing more.

The dependency and CMake subprojects systems share already a lot of code (the entire CMake trace parser for instance), so duplicating cross-compilation/toolchain-file handling logic won't happen.

> Your dependencies already have an option for `cmake_args`.

To be honest, I forgot that this flag existed. It is true that it is technically possible to pass `['-T', '/my/toolchain.cmake']` in `cmake_args`, but this also isn't ideal and will change in the future (with deprecation, etc. of course). I am already working on fixing a similar issue with #7231 for CMake subprojects and may implement something similar for `cmake_args` in the `dependency()` function.

> Again keep the user in control instead of going into auto steer mode and drive into a hardcoded wall

I do want to give users control, but with more explicit options than just one big array that gets appended to the command-line. If users start adding cross files via the command-line, then I will have a hard time implementing proper cross-compilation support. The same goes for any other CMake options (besides `-D`). If there are valid reasons for adding some other flags, I will add support for it. However, the CMake integration is far from trivial, and adding something like `--trace-format=human` will break meson.

So, it is not about adding "Only one option....." (which indeed is trivial), but about the implications for the future meson development.

**Comment by ceztko** [comment]
Apparently I'm having the same issue using meson in Windows when building third party dependencies. Since there's no convention on where headers and libs will be searched, the usefulness of trying to use CMake to find dependencies in Windows is lower. In CMake I do the following to alter the search paths for headers and libs:

```
# Set include folders for find_package()
list(APPEND CMAKE_INCLUDE_PATH
    "${DEPS_PATH}/zlib/include"
    #...
)

# Set library folders for find_package()
list(APPEND CMAKE_LIBRARY_PATH
    "${DEPS_PATH}/zlib/lib/${ARCH}"
    #...
)
```

Where `ARCH` is a variable of mines. So if I could pass `CMAKE_INCLUDE_PATH` and `CMAKE_LIBRARY_PATH` from the command line I could properly use CMake inside meson to find dependencies in Windows, without editing the 3rdparty dependencies meson build files.

**Comment by mensinda** [comment]
Implemented in #7816
